flycheck-rust.el

```diff
@@ -150,7 +150,21 @@ description of the conventional Cargo project layout."
                                             (file-name-directory manifest)))))
             ;; If all else fails, just pick the first target
             (car targets))))
-      (let-alist target (cons (flycheck-rust-normalize-target-kind .kind) .name)))))
+      (let-alist target
```

Comment:
Couldn't we just filter out the `custom-build` target from the `targets` list before?  We should fallback on picking the first one without having to iterate through targets again.

Reply:
Hmm...
I think it'll be more complex if we do so.

Reply:
Ah, I see, because we need to make sure to pick up a target *from the same package*, and the relation between packages is lost by `flycheck-rust-get-cargo-targets`.

I'm worried about the way the target from the same package is picked then:
```
(string-prefix-p
  (expand-file-name "src" (file-name-directory file-name))
  .src_path))) 
```
There is no indication that it is, indeed, part of the same package, as returned by `cargo metadata`.

What about doing some pre-processing at the end of `flycheck-rust-get-cargo-targets` to replace `custom-build` targets by another target from the same package?  This would leave `flycheck-rust-find-cargo-target` untouched, and would have the same result, but with the additional guarantee that the `custom-build` replacement comes from the same package.

Reply:
>There is no indication that it is, indeed, part of the same package, as returned by cargo metadata.

OK.
We have to search targets only from same packages as current buffer.

> What about doing some pre-processing at the end of flycheck-rust-get-cargo-targets to replace custom-build targets by another target from the same package? 

If we do so, how we know a target is from the same package as current `build.rs`?
I think it's natural to process after finding that current target is `custom-build`, because it's a tricky corner case, as you say.

Reply:
> If we do so, how we know if a target is from the same package as current build.rs?

Sorry for being unclear.  In the `build-script-test` package, we have two targets, as returned by `cargo metadata`:
```
 "targets": [
        {
          "kind": [ "bin" ],
          "crate_types": [ "bin" ],
          "name": "build-script-test",
          "src_path": ".../build-script-test/src/main.rs"
        },
        {
          "kind": [ "custom-build" ],
          "crate_types": [ "bin" ],
          "name": "build-script-build",
          "src_path": ".../build-script-test/build.rs"
        }
      ],
```
I was suggesting that, before extracting these targets to pass them to `flycheck-rust-find-cargo-target`, we do the following processing:
```
 "targets": [
        {
          "kind": [ "bin" ],
          "crate_types": [ "bin" ],
          "name": "build-script-test",
          "src_path": ".../build-script-test/src/main.rs"
        },
        {
          "kind": [ "bin" ],
          "crate_types": [ "bin" ],
          "name": "build-script-test",
          "src_path": ".../build-script-test/build.rs"
        }
      ],
```
That is, we change the `kind`, and `name` properties of `custom-build` targets to the values of the first target in the same package.  That way, `flycheck-rust-find-cargo-target` will find a matching target for the `build.rs` file, and use the `kind` and `name` values of the first target.

Reply:
Thanks for explanation.
I really misunderstand your `pre-processing`, and, it looks good.
I'll try it.

Reply:
>  honestly now I feel this implementation more complicated than the current one.

I'll trust you on that.  Glad we could discuss alternatives.